### ISSUE TITLE
Resuming error handling functions after a signal

### ISSUE THREAD

**Issue by bmw** [certbot/certbot#3219]
Despite me writing `certbot/error_handler.py` many months ago, it just surprised me. What surprised me is more or less documented below:

> Every registered function is attempted to be run to completion exactly once. If a registered function raises an exception, it is logged and the next function is called. If a (different) handled signal occurs while calling a registered function, it is attempted to be called again by the next signal handler.

I think not calling a registered function again if it just raised an exception makes sense. If it errored the first time, it's likely to error again. I'm not sure if calling the function again if it is interrupted by a signal, however, is the right behavior. Registered functions can change state and an interrupted function may leave something in an invalid state such that calling it again causes incorrect behavior.

For a real life example, `client.py` registers the method `_restart_and_rollback` with the `ErrorHandler`. This method rolls back one checkpoint and restarts the plugin. What if, however, a signal interrupts the method after rolling back the checkpoint. Calling it again will attempt to rollback another checkpoint which is incorrect.

I think we should call try and call a registered function at most once to avoid this problem.

**Comment by pde** [comment]
Hmmm. This is a little subtle and weird. If it's a control-c signal which lands early in the function's execution, you could argue that retrying the function is correct.  But you're right that that (1) creates a danger of double rollbacks and (2) will often attempt a futile action twice.

I wonder if there's any sane way to get to a certain point and say "do not retry after here"?  Or is that too complicated?

**Comment by pde** [comment]
A menu of strategies for this:

https://stackoverflow.com/questions/842557/how-to-prevent-a-block-of-code-from-being-interrupted-by-keyboardinterrupt-in-py

**Comment by bmw** [comment]
Yeah I saw that after we talked in the hallway. As specified, none of them work for us due to one reason or another, however, it helps suggest strategies for how it could work.

Proposal that was resolve this and #3215:

At the start of error handling, set every signal handler we care about to one that simply records the signal that was raised. After calling all cleanup functions, we restore the original signal handlers and call each of them in the order they occurred.

Potential downsides of this approach:

1) A new signal will interrupt us calling the signal handlers of the deferred signals. I don't think we really care about this because this is the behavior you get by default in Python -- a new signal, including one of the same type, interrupts a signal handler.
2) We don't pass the stack frame from when the signal was first encountered to the signal handler. I'm not aware of a way to reliably do this ([getsignal](https://docs.python.org/2/library/signal.html#signal.getsignal) doesn't always return a callable and often doesn't if the signal handler is the default). Again, I'm not sure we really care and passing the current stack frame is arguably more correct.
